include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h

```diff
@@ -0,0 +1,681 @@
+{
+namespace __par_backend_hetero
+{
+
+// Baseline wrapper which provides no acceleration via SLM memory, but still
+// allows generic calls to a wrapped binhash structure from within the kernels
+template <typename _BinHash, typename _ExtraMemAccessor>
+struct __binhash_SLM_wrapper
+{
+    _BinHash __bin_hash;
+    __binhash_SLM_wrapper(_BinHash __bin_hash_, _ExtraMemAccessor /*__slm_mem_*/,
+                          const sycl::nd_item<1>& /*__self_item*/)
+        : __bin_hash(__bin_hash_)
+    {
+    }
+
+    template <typename _T>
+    ::std::uint32_t
+    get_bin(_T&& __value) const
+    {
+        return __bin_hash.get_bin(::std::forward<_T>(__value));
+    }
+
+    template <typename _T>
+    bool
+    is_valid(_T&& __value) const
+    {
+        return __bin_hash.is_valid(::std::forward<_T>(__value));
+    }
+};
+
+// Specialization for custom range binhash function which stores boundary data
+// into SLM for quick repeated usage
+template <typename _Range, typename _ExtraMemAccessor>
+struct __binhash_SLM_wrapper<oneapi::dpl::__internal::__custom_range_binhash<_Range>, _ExtraMemAccessor>
+{
+    using _bin_hash_type = typename oneapi::dpl::__internal::__custom_range_binhash<_Range>;
+
+    _bin_hash_type __bin_hash;
+    _ExtraMemAccessor __slm_mem;
+    __binhash_SLM_wrapper(_bin_hash_type __bin_hash_, _ExtraMemAccessor __slm_mem_, const sycl::nd_item<1>& __self_item)
+        : __bin_hash(__bin_hash_), __slm_mem(__slm_mem_)
+    {
+        //initialize __slm_memory
+        ::std::uint32_t __gSize = __self_item.get_local_range()[0];
+        ::std::uint32_t __self_lidx = __self_item.get_local_id(0);
+        ::std::uint8_t __factor = oneapi::dpl::__internal::__dpl_ceiling_div(__bin_hash.__boundaries.size(), __gSize);
+        ::std::uint8_t __k = 0;
+        for (; __k < __factor - 1; ++__k)
+        {
+            __slm_mem[__gSize * __k + __self_lidx] = __bin_hash.__boundaries[__gSize * __k + __self_lidx];
+        }
+        // residual
+        if (__gSize * __k + __self_lidx < __bin_hash.__boundaries.size())
+        {
+            __slm_mem[__gSize * __k + __self_lidx] = __bin_hash.__boundaries[__gSize * __k + __self_lidx];
+        }
+    }
+
+    template <typename _T>
+    ::std::uint32_t
+    get_bin(_T&& __value) const
+    {
+        return _bin_hash_type::get_bin_helper(__slm_mem.begin(), __slm_mem.end(), ::std::forward<_T>(__value));
+    }
+
+    template <typename _T>
+    bool
+    is_valid(_T&& __value) const
+    {
+        return _bin_hash_type::is_valid_helper(__slm_mem[0], __slm_mem[__slm_mem.size() - 1],
+                                               ::std::forward<_T>(__value));
+    }
+};
+
+template <typename _BinHash, typename _ExtraMemAccessor>
+auto
+__make_SLM_binhash(_BinHash __bin_hash, _ExtraMemAccessor __slm_mem, const sycl::nd_item<1>& __self_item)
+{
+    return __binhash_SLM_wrapper(__bin_hash, __slm_mem, __self_item);
+}
+
+template <typename _BinHash>
+struct __binhash_manager_base
```

Comment:
I think expressing of the polymorphic behavior is so "complicated" and tricky here. 
We have a "base" class (which does nothing)  with a couple of dummy methods and a derived class which inherits just  `_BinHash ` field.
But by fact we have two class specializations. 
I believe the code may be reduced and made more shorter and cleared.

We have already two specializations  - `__evenly_divided_binhash` and `__custom_range_binhash`...
We may add a general bool property, for example, to "inform" kernel whether  we need SLM or don't.

Reply:
I experimented with this using template specialization, and was unhappy with the result, and went back to the polymorphic approach.  You can see the difference between that attempt and the current PR at
https://github.com/oneapi-src/oneDPL/pull/1273/commits/efbb843feebe50f2f44ab79752ef9d80939174a0 .  I'm sure there is a better way to do it with the template specialization way, without having to explicitly specify the types, but I'm not finding it.

I think I prefer the polymorphic approach because we *are* building functionality on, by making changes to accommodate extra memory that needs a call to `require_access()`, has a size for SLM allocations, and requires a `sycl::buffer` lifetime extender.  Both need to return a valid bin hash function to be used in the kernel.

As far as the bool property for SLM vs no SLM (I assume instead of the zero SLM memory return), we should discuss this further offline.  I would prefer the kernels not need to know the details of the binhash function and be written generically.  I think this makes the kernels cleaner, and doesn't need branching to special case for individual binhash functions.  I'm open to a discussion but there is some nuance that needs to be worked through with code examples.